### PR TITLE
Add container id support to Resource

### LINKED ISSUE #1981: Container Resource MetaData auto detection in Resource SDK
### Problem Statement

In the spirit of what was done in #1788 #1840 for Process and Operating system auto detection - It is worthwhile to have an auto detection processor that collects and reports container information `container.id` and ~`container.name`.~

### Proposed Solution

Most of the container runtime environments (Docker, CRI-O, and containerd) have this informaton in cgroup files on the container filesystem. Resource SDK initialization goes through these files and fetches it. 

#### Alternatives

Other option is to set environment variables on containers. But that is still sort of manual thing to do or some orchestration agent does it (like K8, PAAS environement)    

cc: @stevebappd

### PR DESCRIPTION
**Pull request by XSAM**
resolves #1981

## Prior Art

- https://github.com/open-telemetry/opentelemetry-java/pull/3321

cc: @pavankrish123

### CODE REVIEW

**Review by MadVikingGod — APPROVED** [review]

**Comment by jcchavezs** [comment]
I was checking this PR, I was wondering if the approach I took in https://github.com/hypertrace/goagent/blob/main/sdk/internal/container/container.go was simpler or am I missing something. I also wonder, do we want an integration test like https://github.com/hypertrace/goagent/tree/main/tests/docker. I could port it here @XSAM

**Comment by XSAM** [comment]
> I was wondering if the approach I took in https://github.com/hypertrace/goagent/blob/main/sdk/internal/container/container.go was simpler or am I missing something.

This test case shows a cgroup line may contain the prefix and the suffix.
https://github.com/open-telemetry/opentelemetry-go/blob/bf3c93cb81f262e77b9839443d1c46395fc48466/sdk/resource/container_test.go#L49-L51

Also, here is an example of the cgroup string that is fetched from my Kubernetes cluster. You can see the cgroup string becomes complicated.

```
2:cpuset:/kubepods/besteffort/podbd31ef54-609d-44dd-85f8-12014a486358/a4a47304716dc887fc6c8e389bfd32c0f98f3571d41d220e43dbc6d0eec67182
```

> I also wonder, do we want an integration test like https://github.com/hypertrace/goagent/tree/main/tests/docker.

Good point. I also wonder about this. To achieve this integration test, we need container runtime. It means to run the tests, it requires all contributors to install a container runtime. It also increases the dependence on this project. Not sure it is worth doing that. Any thoughts? @open-telemetry/proto-go-approvers

**Comment by XSAM** [comment]
BTW, I found the approach of this PR is not working for cgroup v2. In cgroup v2, we only get little info from the cgroup file.

For instance,

```
0::/
```

File: sdk/resource/container.go
Comment (seh) [inline sdk/resource/container.go:36]: Given that this detector uses a specific technique to find this container's ID—this one assuming that one extraced from the cgroup file is the right choice—would you be amenable to naming this detector more specifically? Perhaps `cgroupContainerIDDetector`? That would clarify where this ID is coming from, and open us to other approaches to finding a container's ID.

I'm not sure whether it would warrant a corresponding change to the proposed `WithContainerID` function.

**Comment by jcchavezs** [comment]
I am curious about this. Instead of opening up an API to this I would focus
all efforts to get a unique strong implementation. Are you targeting a
concrete case?

**Comment by seh** [comment]
> Are you targeting a concrete case?

Are you asking me? If so, I'm telegraphing the conversation from today's Go SIG meeting—a conversation that I started. There will be other interpretation of what "a container's ID" is or what it means. I don't know what they are. I know enough to doubt the universality of this particular choice.

**Comment by MadVikingGod** [comment]
So I approved this with my understanding at the time that both cgroups were made in a response to containers, and that the current Linux runtimes, that I know of, all use cgroups under the hood to manage containers.

While both the OCI, and the CRI require a container ID, neither specifies that you must use cgroups to do so.  I don't know if there is an implementation counter to this, but I will admit I haven't checked all of them.

With that in mind, these would be my recommendations:
- Remove `defaultContainerIDProvider`.  You can still have this setup if needed for testing, but it shouldn't be `default`, instead, there should be the option for many different detectors.
- `containerIDDetector` should have cgroup in the name.  It uses cgroups to find the containerID

**Comment by XSAM** [comment]
FYI, I created [an issue](https://github.com/opencontainers/runtime-spec/issues/1105#issuecomment-1002888071) on OCI spec looking for an official/stable way to fetch container id within the container.

**Comment by XSAM** [comment]
> looking for an official/stable way to fetch container id within the container.

This should be a long-term direction and not block this PR.

**Comment by XSAM** [comment]
I believe this PR is ready for review.

File: sdk/resource/container.go
Comment (Aneurysm9) [inline sdk/resource/container.go:114]: Does this boil down to this regex?  Would it be better to use the regex instead?  It took me a while to figure out what the code was trying to match.

https://regex101.com/r/xlx8Dv/1 shows it to be equivalent.

```
^.*\/(?:.*\-)?([0-9a-f]+)(?:\.|\s*$)
```

File: sdk/resource/process_test.go
Comment (Aneurysm9) [inline sdk/resource/process_test.go:112]: Why these changes?  They seem unrelated.

Comment (XSAM) [reply sdk/resource/process_test.go:112]: Yeah, these are unrelated typos. I can fix these in another PR if you like.

File: sdk/resource/container.go
Comment (XSAM) [reply sdk/resource/container.go]: Sure 👍

**Review by dmathieu — APPROVED** [review]

File: sdk/resource/container.go
Comment (hanyuancheung) [inline sdk/resource/container.go:100]:
```suggestion
	if len(matches) <= 1 {
		return ""
	}
	return matches[1]
```
I prefer this could be better. A little suggestion ~

File: sdk/resource/container.go
Comment (hanyuancheung) [inline sdk/resource/container.go:38]:
```suggestion
var (
	containerID containerIDProvider = getContainerIDFromCGroup
	cgroupContainerIDRe = regexp.MustCompile(`^.*/(?:.*-)?([0-9a-f]+)(?:\.|\s*$)`)
)

type cgroupContainerIDDetector struct{}

const cgroupPath = "/proc/self/cgroup"
```

File: sdk/resource/resource_test.go
Comment (hanyuancheung) [inline sdk/resource/resource_test.go:661]:
```suggestion
```
Remove blank line

**Review by hanyuancheung — APPROVED** [review]
lgtm!

**Review by Aneurysm9 — APPROVED** [review]

**Pull request merged**
